crates/sui-core/src/state_accumulator/mod.rs

```diff
@@ -0,0 +1,394 @@
+                        self.accumulate(state)
+                            .await
+                            .expect("Accumulation failed");
+                    }
+                }
+                // schedule new tasks
+                state = self.queue.recv() => match state {
+                    Some(state) => {
+                        debug!(
+                            ?state,
+                            "StateAccumulator: received enqueued state",
+                        );
+
+                        let store = self.store.clone();
+                        let state_clone = state.clone();
+
+                        pending.push_back(spawn_monitored_task!(async move {
+                            let accumulator = hash_checkpoint_effects(state_clone.clone(), store)?;
+                            Ok((accumulator, state_clone))
+                        }));
+                    }
+                    None => panic!("StateAccumulator: all senders have been dropped"),
+                }
+            }
+        }
+    }
+
+    /// Unions all checkpoint accumulators at the end of the epoch to generate the
+    /// root state hash and saves it. This function is guaranteed to be idempotent (despite the
+    /// underlying data structure not being) as long as it is not called in a multi-threaded
+    /// context.
+    async fn accumulate(&self, last_state_of_epoch: State) -> Result<(), TypedStoreError> {
+        let State {
+            epoch,
+            checkpoint_seq_num: last_checkpoint,
+            ..
+        } = last_state_of_epoch;
+
+        if self
+            .store
+            .tables
+            .root_state_hash_by_epoch
+            .contains_key(&epoch)?
+        {
+            return Ok(());
+        }
+
+        let (next_to_accumulate, mut root_state_hash) = self
+            .store
+            .tables
+            .root_state_hash_by_checkpoint
+            .iter()
+            .skip_to_last()
+            .next()
+            .map(|(highest, hash)| (highest.saturating_add(1), hash))
+            .unwrap_or((0, Accumulator::default()));
+
+        for i in next_to_accumulate..=last_checkpoint {
+            let acc = self
+                .store
+                .tables
+                .state_hash_by_checkpoint
+                .get(&i)?
+                .unwrap_or_else(|| {
+                    panic!("Accumulator for checkpoint sequence number {i:?} not present in store")
+                });
+            root_state_hash.union(&acc);
+        }
+
+        // We want to enforce that this table is append only, otherwise we
+        // may end up re-accumulating the same state
+        assert!(
+            !self
+                .store
+                .tables
+                .root_state_hash_by_checkpoint
+                .contains_key(&last_checkpoint)?,
+            "StateAccumulator: root state hash already exists for checkpoint {last_checkpoint:?}"
```

Comment:
we can always insert, because we will always compute the same value to insert. If we want an `assert` here we can assert that if there already was a value, it is equal to the one we are trying to insert.